--- a/heartbeat/ocf-shellfuncs.in	2025-09-29 14:01:55.762931795 +0200
+++ b/heartbeat/ocf-shellfuncs.in	2025-09-29 14:09:28.651731793 +0200
@@ -1093,6 +1093,16 @@
 	echo $1
 }
 
+ocf_promotion_score() {
+	ocf_version_cmp "$OCF_RESKEY_crm_feature_set" "3.10.0"
+	res=$?
+	if [ $res -eq 2 ] || [ $res -eq 1 ] || ! have_binary "crm_master"; then
+		${HA_SBIN_DIR}/crm_attribute -p ${OCF_RESOURCE_INSTANCE} $@
+	else
+		${HA_SBIN_DIR}/crm_master -l reboot $@
+	fi
+}
+
 __ocf_set_defaults "$@"
 
 : ${OCF_TRACE_RA:=$OCF_RESKEY_trace_ra}
