--- a/heartbeat/powervs-move-ip.in	2025-09-15 16:13:34.225046827 +0200
+++ b/heartbeat/powervs-move-ip.in	2025-09-15 17:39:02.746258434 +0200
@@ -33,9 +33,13 @@
 from pathlib import Path
 from urllib.parse import urlparse
 
-import requests
-import requests.adapters
-import urllib3.util
+try:
+    sys.path.insert(0, '/usr/lib/fence-agents/support/ibm')
+    import requests
+    import requests.adapters
+    import urllib3.util
+except ImportError:
+    pass
 
 # Constants
 OCF_FUNCTIONS_DIR = os.environ.get(
