From f75a1bb59fe729b9de65fc3c4436e9edaf8f8f41 Mon Sep 17 00:00:00 2001
From: Oyvind Albrigtsen <oalbrigt@redhat.com>
Date: Wed, 30 Apr 2025 16:09:50 +0200
Subject: [PATCH] powervs-subnet: fix plaintext token file support

---
 heartbeat/powervs-subnet.in | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/heartbeat/powervs-subnet.in b/heartbeat/powervs-subnet.in
index b8f3864e9a..a90bc73c6b 100755
--- a/heartbeat/powervs-subnet.in
+++ b/heartbeat/powervs-subnet.in
@@ -378,8 +378,9 @@ class PowerCloudAPI:
                         api_key = keys.get("Apikey", "")
                         if not api_key:
                             api_key = keys.get("apikey", "")
+                    # data is text, return as is
                     except ValueError:
-                        # data is text, return as is
+                        f.seek(0)
                         api_key = f.read().strip()
             except FileNotFoundError:
                 raise PowerCloudAPIError(
