From 1e546b85010e5fdbf7a0f31207dce144c14c50ec Mon Sep 17 00:00:00 2001
From: Oyvind Albrigtsen <oalbrigt@redhat.com>
Date: Wed, 29 Oct 2025 15:17:30 +0100
Subject: [PATCH] MailTo: add s-nail support for multiple recipients

---
 heartbeat/MailTo | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

diff --git a/heartbeat/MailTo b/heartbeat/MailTo
index 56940bafaa..a3ee6a04c8 100755
--- a/heartbeat/MailTo
+++ b/heartbeat/MailTo
@@ -92,12 +92,16 @@ END
 }
 
 MailProgram() {
-	$MAILCMD -s "$1" "$email"  <<EOF
-        $Subject
- 
-        Command line was:
-        $ARGS
-EOF
+	local body="\
+$Subject
+
+Command line was:
+$ARGS"
+	if $MAILCMD -V | grep -q "^s-nail"; then
+		printf "$body" | $MAILCMD -s "$1" $(echo $email | sed "s/,\s*/ /g")
+	else
+		printf "$body" | $MAILCMD -s "$1" "$email"
+	fi
 	return $?
 }
 
